Fix BaseDocument._from_son() deepcopy failed with compiled regex

StoneMoe opened the pull request:
Fix for issue https://github.com/MongoEngine/mongoengine/issues/2429

bagerard commented on mongoengine/fields.py:
```diff
@@ -104,8 +105,24 @@ def __init__(self, regex=None, max_length=None, min_length=None, **kwargs):
         self.regex = re.compile(regex) if regex else None
         self.max_length = max_length
         self.min_length = min_length
+        self._kwargs = kwargs
```
I don't think this line is useful

StoneMoe replied:
This line is for another abandoned approach which I should clean it up, thanks.

bagerard commented on mongoengine/fields.py:
```diff
@@ -104,8 +105,24 @@ def __init__(self, regex=None, max_length=None, min_length=None, **kwargs):
         super().__init__(**kwargs)
 
+    def __deepcopy__(self, memo):
+        memo[id(self)] = self
```
I believe this is already done in the call `copied_obj = copy.deepcopy(self, memo)` a few line below

StoneMoe replied:
Yes...I found I got it wrong after some digging.
`memo[id(self)] = self` actually make `deepcopy(self)` bypass deepcopy operate totally.

bagerard commented on mongoengine/fields.py:
```diff
@@ -104,8 +105,24 @@ def __init__(self, regex=None, max_length=None, min_length=None, **kwargs):
+        regex_bak = self.regex
+        self.regex = None
+
+        copied_obj = copy.deepcopy(self, memo)
```
isn't this causing recursion to this same method?

StoneMoe replied:
I'll try another approach. since `memo[id(self)] = self` is not a real fix for recursion issue

bagerard commented on tests/document/test_instance.py:
```diff
@@ -3911,6 +3911,13 @@ def test_dbfields_are_loaded_to_the_right_modelfield_for_dyn_doc_2(self):
+            re_str = StringField(regex=r"(^ABC\d\d\d\d$)")
+
+        data = {"re_str": "ABC0000"}
+        User._from_son(son=data, _auto_dereference=False)
```
I think it would be better to push the test further and do additional assertions on the returned User instance, for instance check that `new_user.regex is not User.regex`.

StoneMoe replied:
Since compiled regex objects are atomic, I'll only check regex pattern is same string and StringField is not same object.

bagerard commented on tests/document/test_instance.py:
```diff
@@ -3911,6 +3911,13 @@ def test_dbfields_are_loaded_to_the_right_modelfield_for_dyn_doc_2(self):
+        data = {"re_str": "ABC0000"}
+        User._from_son(son=data, _auto_dereference=False)
+
 
```
I'm not used to overwriting deepcopy but from what I read, there are known risks of recursion problem when overwriting deepcopy, so it would be great to have a test that covers that and make sure it doesn't loop infinitely. Could you also add another test with a User having a a reference to itself, something like

```
import copy

class User(Document):
    name = StringField(regex=r"(.*)")
    other_user = ReferenceField('self')

user1 = User(name='John').save()
user2 = User(name='Bob', other_user=user1).save()

user1.other_user = user1
user1.save()

for us in User.objects():
    copied_user = copy.deepcopy(us)
    assert copied_user is not us
    assert copied_user.regex is not us.regex
```

Since that new test would relate to StringField more than instance, that one would make sense to be in test_string_field.

StoneMoe replied:
Sure, I'll try to add more test.

bagerard reviewed (changes requested):
Hi @StoneMoe , thanks for your contribution. It looks great, I only have a minor few comments. Please check the comments and let me know if there are additional questions.

StoneMoe commented:
I have no idea why `deepcopy()` works for the Document instance `_fields` attribute returned by `_from_son()`, but not work for the objects returned from queryset.....

StoneMoe commented:
Hi, thanks for the reviewing, I added more tests and changed approach for this fix based on that.

but there is a question might need more discussion:
1. `deepcopy()` seems won't touch the attributes added by __new__, which is `instance._fields` in this case.
2. which resulting in the test you request to add won't pass in the first place, is this expected or another bug for mongoengine?
3. `deepcopy()` works for `_fields` attribute of the instance returned by `_from_son()`, but not work for the objects returned from queryset

bagerard commented:
sorry I lost track of this. I tested the initial issues on latest master and it's not failing anymore so I guess something got merged since this PR got created. I'm including the test cases in https://github.com/MongoEngine/mongoengine/pull/2865 to avoid regression but I'm closing this one as I prefer to avoid too much hacks on deepcopy internals unless absolutly necessary
